Clear Inventory Chatcommand (/clearinv)

octacian opened the pull request:
This PR introduces a chatcommand which allows players to clear either their own inventory or another player's inventory. `/clearinv` with not parameters clears the inventory of the player who executed the command, while `/clearinv <playername>` attempts to clear the inventory of the player specified.

Clearing your own inventory requires no additional privileges, while clearing another player's inventory requires the pre-existing `server` privilege.

I've tested both in singleplayer and multiplayer.

rubenwardy commented on builtin/game/privileges.lua:
```diff
@@ -90,3 +90,7 @@ core.register_privilege("debug", {
 	give_to_singleplayer = false,
 })
+core.register_privilege("clearinv", {
+	description = "Can use clear inventory command",
```
for other players

Also, maybe rename to make it even more obvious it's not for your own inv

octacian replied:
Maybe just require `server` priv... IDK. For now, I'll change desc and leave priv name until better suggestion.

rubenwardy commented on builtin/game/chatcommands.lua:
```diff
@@ -939,3 +939,34 @@ core.register_chatcommand("last-login", {
+				return false, "Player must be online to clear inventory!"
+			end
+		else
+			local player = core.get_player_by_name(name)
```
code share possible here:

```lua
local player
if param and param ~= "" then
    -- check priv, set player to player
else
   -- set player to caller
end

if player then
    -- clear inv of player
else
    -- error
end
```

rubenwardy commented on builtin/game/chatcommands.lua:
```diff
@@ -939,3 +939,34 @@ core.register_chatcommand("last-login", {
+		else
+			local player = core.get_player_by_name(name)
+			if player then
+				player:get_inventory():set_list("main", {})
```
craft? craft result?

octacian commented:
@rubenwardy will amend when I change a few things.

octacian commented:
As @rubenwardy requested:
* updated priv desc
* simplified code for chat command
* made command clear craft and craftpreview inventories

I do agree that the privilege needs a new name altogether, but I'm really not sure what to call it. I was thinking `clear_others_inv` as I stated in an above comment to rubenwardy's review, but I'd like some extra input on this.

SmallJoker commented:
Apart from creative mode - does this command have any use case?

nerzhul commented:
@SmallJoker it's useful for modders or developers to clear their inventory sometimes

rubenwardy commented on builtin/game/chatcommands.lua:
```diff
@@ -939,3 +939,32 @@ core.register_chatcommand("last-login", {
+			player:get_inventory():set_list("main", {})
+			player:get_inventory():set_list("craft", {})
+			player:get_inventory():set_list("craftpreview", {})
+			if pmsg then core.chat_send_player(param, pmsg) end
```
Don't use one line if statements

octacian replied:
@rubenwardy fixed. I'll keep that in mind in the future :D

rubenwardy commented on builtin/game/chatcommands.lua:
```diff
@@ -939,3 +939,34 @@ core.register_chatcommand("last-login", {
+	func = function(name, param)
+		local player, msg, pmsg
+		if param and param ~= "" then
+			if not core.check_player_privs(name, {clearinv=true}) and name ~= param then
```
maybe move `name ~= param` to the if statement on line 948?
Stops the `cleared your inventory` being printed it it's yourself

rubenwardy commented on builtin/game/chatcommands.lua:
```diff
@@ -939,3 +939,34 @@ core.register_chatcommand("last-login", {
+	params = "[name]",
+	description = "Clear the inventory of yourself or another player",
+	func = function(name, param)
+		local player, msg, pmsg
```
msg is unused

octacian commented:
Made changes requested by rubenwardy.

BTW, the `msg` variable had been used to return a different value depended on whether the player's own inventory or that of another player was being cleared. But, it is true that I'm not using it anymore as I decided it wasn't really needed, so removed.

bigfoot547 commented:
You may borrow some code from (here)[https://github.com/bigfoot547/morecommands/blob/master/init.lua#L303-LL355]

rubenwardy commented on builtin/game/chatcommands.lua:
```diff
@@ -939,3 +939,34 @@ core.register_chatcommand("last-login", {
+						.. " to run this command (missing privilege: clearinv)"
+			end
+			player = core.get_player_by_name(param)
+			pmsg = name.." cleared your inventory."
```
Maybe send the message here? Removes need for pmsg

octacian replied:
@rubenwardy done. Not sure why I didn't do that in the first place.

bigfoot547 commented:
Hello @octacian , I have some suggestions :smile:!
 - `[maxcount]` argument to remove only this many _items_. (Not itemstacks, ItemStack:get_count())
 - `[item]` argument to only remove a certain itemstring.
 - `[listname]` only remove from this invref list.
 - This is somewhat in the noise, but after execution, tell the player how many ItemStacks and items were removed.

Thanks bigfoot547.
Also :+1:.

octacian commented:
@bigfoot547 interesting suggestions. First thing, of most importance, why should I not use `inv:set_list`? Iterating through each itemstack seems incredibly inefficient. About the other things, I think that they should be a part of a different command, maybe `/clearstack` not `/clearinv`. IDK, just a thought. Adding all that to this command just seems a little overly complicated considering how it would generally be used. I really understand why being able to clear only specific things would be useful, just don't think it fits in this PR.

bigfoot547 commented:
Okay. Thanks for the consideration :wink:

sofar commented:
What exactly requires this to be in `builtin` ?

In other words, please explain why this can't be put inside `minetest_game` or `minimal_game` instead, and why this command is needed for every (**every**) subgame that exists.

Once we add this, it will be in every subgame. This will make it more difficult for subgames that rely on items not being destroyable, and may be abused by players. Imagine for instance a CTF game where someone grabs the opponents flag and does `/clearinv` making it impossible for his team to win.

Solutions: 
- put this in `minetest_game` instead?
- require a priv to empty your own inventory? (not my preference)

Note: the fact that a trashcan exists in some of the inventory mods is not relevant: mods/subgames can remove the trashcan as well as the possibility to drop items or prevent them from being destroyed in e.g. lava.

octacian commented:
This doesn't need to be in Minetest game. Other devs said I should put it in builtin, that's the only reason why it's here. If it were to be put in MTG, I really think that rather than creating a new mod like killme, we should just add a specifically for trivial chatcommands like this. (I think that killme should be moved to something like that either way.)

If `/clearinv` stays in builtin, I agree that a priv for players to clear their own inventory isn't the way to go. I think we'd be better off to simply introduce a setting (maybe `disable_clearinv`) that if `false` would prevent the command and privilege from being registered. Not sure if that's a good way to do it at all, just throwing around ideas.

I can see, though, that putting it in MTG could still be preferable because subgames that didn't need the functionality wouldn't need to make any changes. I'm more than willing to open a PR on MTG instead.

sofar commented:
A setting is client-side, so it would have no effect.

Note: I'm perfectly OK with having this command widely available, so I'm puzzled how we can nicely solve this issue.

Maybe we should just assume the `server` priv is required for clearing out other players privs, and add a new (default) priv allowing players to empty their own inventory? Then servers can revoke that priv for all players.

kahrl commented:
Players can already delete items with the `/pulverize` command, which is also in builtin and requires no privs.

Maybe an API could be provided that allows mods to delete already registered chat commands?
```
# in builtin/game/chatcommands.lua
function core.unregister_chatcommand(cmd)
    core.chatcommands[cmd] = nil
end
```

octacian commented:
@sofar just went on a CTF as I thought the flag wasn't stored as an inventory item. It isn't. So, `/pulverize` or `/clearinv` shouldn't create any gameplay issues for CTF.

Either way, I'd support a feature to either unregister or override a command. Might be nice to have both. I'd be happy to submit a PR for unregistering and/or overriding.

bigfoot547 commented:
What happens when a mod creates its own inv list for the player?

`unified_inventory`, `3d_armor comes to mind`.

octacian commented:
Since #5076 has been merged, I'd say this is all ready to go. With the ability to override and unregister chatcommands, the issue with mods/games that don't want this should be dealt with.

@bigfoot547 I think that mods like unified_inv or 3d_armor should either provide their own command or override this one. I guess I could iterate through the lists, but then what if there was a mod that _didn't_ want all the inventories cleared. (e.g. clearing all would cause the player hand to be reset to default.)

Zeno- commented:
Added WIP
See discussion at: http://irc.minetest.net/minetest-dev/2017-01-26#i_4795915

paramat commented:
> remove the line to clear crafting grid

Yes please.

twoelk commented:
I think the option _"Can clear another player's inventory"_ would come in extremely handy in an educational environment. In such a use case this would not only be usefull for the admin but rather any staff or moderator that might be assigned the job only temporarly. 

Think of a class that visits an existing world to take up a new task. the teacher could assign one pupil as moderator who may need to clear the inventories of all other players now and then to get the task done as planned out to achieve some educational goal. Just as he might need the ability to provide the pupils in bulk or predefined groups with certain amounts of resources. 

I think minetest indeed lacks some more tools a teacher or an admin might find useful to monitor players in groups. As interest seems to grow for minetest as educational tool I do hope development would listen more to the needs of educational personall as I could imagine minetest would benefit greatly from a gain in popularity among such people and institutions.

Actually it might indeed be extremely usefull if minetest had a built in function to assign players to more groups that may even overlap and commands could be targeted at certain groups of players be it based on privs or maybe proximity to some coordinates, special nodes or even elements within the player name. 

Of course anything as powerfull as this should be completely configurable, so a server admin or a game designer may decide to use the feature, hide it or completely dissable it. Hoping #5076 catches all use cases.

octacian commented:
With twoelk's comment above, I realize that in some cases that he mentioned above, one might want the craft grid to be cleared as well. Otherwise, a student could stash something there that they were not supposed to have. And, if a student in each group were to be assigned to be a "moderator," you wouldn't want to give them the `server` priv.

I won't move to the server priv right away, I'd like some more feedback. And, with the craftgrid, I think the best was to go is to either introduce a command like `clearcraft` or an option like `clearinv_clears_craftgrid`. I think the best way would be a command specifically for it, will need some feedback though.

paramat commented:
How about keeping it simple, a single command to clear inventory, crafting grid and output?

octacian commented:
Personally, I am against merging without the ability to clear other's inventories. Yes, the classroom example is narrow, but it is important and isn't the only place features like that could be needed. Take an adventure map, for example, the creator might want to clear the player's inventory.

Yes, the extra privilege is not necessarily needed, though it _could_, be useful in the classroom or other situations. However, considering that there is an API to override chat commands it shouldn't really be an issue as if used in a school, I would assume that they would already have a mod in use to make MT work better for their school. However, I'm still neutral on this. On one hand, it's nice for things to just work, rather than having to make modifications. Really, though, the extra privilege is like 3 or 4 lines... IDK

About clearing both the inventory and crafting grid. This too could be overridden, so might not be an issue. The original reason for not clearing the crafting grid was because one might want to have a place to put items that they didn't want to lose (e.g. screwdriver). But now that I think about it, that's what chests are supposed to be for.

rubenwardy commented on builtin/game/chatcommands.lua:
```diff
@@ -939,3 +939,31 @@ core.register_chatcommand("last-login", {
+	params = "[name]",
+	description = "Clear the inventory of yourself or another player",
+	func = function(name, param)
+		local player, pmsg
```
pmsg is unused

octacian replied:
Fixed. Missed it in last update somehow.

paramat commented:
The priv has the same name as the basic command, to avoid confusion perhaps it should be clearotherinv, clearinvother or something that makes clear it is for clearing other player's inventories?
I still disagree with adding a priv and ability to clear other player's inventories :-1: it will only be very rarely used or needed.

paramat commented:
I disagree with adding the ability to clear other player's inventories, such a feature is so bizarre and specialist i find it amusing. I know a use for that can be imagined, but that's not enough justification, it's not something that will be often useful for many players.
:-1: 
Being able to clear your own inventory is useful for many players though i will not oppose this PR if only that is added.
COncerning where to put this, in MTGame or builtin, i'm neutral.

Wuzzy2 commented:
I approve. :-)

Hmm, after 10 seconds of thinking I have the ultimate undisputable use case for clearing other people's inventories: Command blocks (from Mesecons)! And placing them in adventure maps.

But in either case, I approve no matter if this command can only clear your own inventory or the inventory of any player.

paramat commented:
Ok, more people are saying this is useful, and this now uses the server priv, so i'm neutral now.

sorcerykid commented:
I seem to recall another discussion in the issue tracker how privileges are "supposed to describe abilities not status" and that "one priv for many different things is not nice" and that "server owners want to grant only specific abilities." Clearing player inventories has nothing to do with server operations. Hence there should be a specific privilege for clearing player inventories.

octacian commented:
@sorcerykid that's true, the `server` privilege is kinda an exception, though I personally still prefer to introduce new privileges for stuff like this it's really up to the devs.

Anybody got further thoughts?

octacian commented:
From my perspective, this is ready to merge. All of the requested changes have been made, and to my knowledge, I've satisfied those who have commented.

Anyone else willing to review? I'd really like to see this merged before 0.4.16.

paramat commented:
First post and commit message needs updating about required priv.

octacian commented:
@paramat Done.
